### PR TITLE
Active sampling simplified

### PR DESCRIPTION
**Pull request by mnoukhov**
all integration tests passing:
- [single](https://beaker.allen.ai/orgs/ai2/workspaces/open-instruct-dev/work/01K9J854NGRQK26277179C9TG8?taskId=01K9J854NN64GXSE339WDF7PMS&jobId=01K9J854S52NCTF90SYXMDSGJ2) 
- [large scale ](https://beaker.allen.ai/orgs/ai2/workspaces/open-instruct-dev/work/01K9G2D4TEDZ00CZ896JG75NSV?taskId=01K9G2D4THRWVE3MPZK59PD6BK&jobId=01K9G2D4Y1KXFSS1R5TKN6M0V8)
- [32B smoke](https://beaker.allen.ai/orgs/ai2/workspaces/open-instruct-dev/work/01K9J97RKSFKRG1XNVG70V78MB?taskId=01K9J97RKYT4SH64PAC7N88XGQ&jobId=01K9J97RQCS26YMFC5SWFVNKTX)

true dynamic sampling aka active sampling: we can filter out 0 std prompts and continuously sample until we have a full batch_size number of non-zero-std prompts

major changes:
- moves reward function into `accumulate_inference_batches` and simplifies the workflow while being better async 
- active sampling now always waits to collect the exact batch size, no `max_retries`

minor changes
- we're now calling reward for every sample so I've made the Timer logging it a `noop` and the reward logging a `debug` so that it doesn't clutter the logs. 
- extended the `Batch` dataclass to include all the fields necessary for a training batch. Can alternatively change to have a new dataclass that's just exactly the fields we need for a training batch and change current `Batch` to be a `GenerationBatch`
- `episode` now refers to a `training episode` (number of samples we train on) instead of a `generation episode` (number of samples we've generated) since everything is running completely async. We can move it back to a good approx of a `generation episode` but different runs will no longer have synced episode numbers for steps / updates  
- `reward_fn` does not take `Batch` anymore, which didn't make much sense as we only used `ground_truths` and `datasets`.  Instead we match `ppo.py` in explicitly passing `ground_truths` and `datasets` 

---

> [!NOTE]
> Adds active sampling with per-sample reward computed during accumulation, extends Batch and metrics, updates training/eval flow, tests, and scripts.
> 
> - **Core RL changes**:
>   - Introduce `Args.active_sampling` (replaces `fill_completions`) to keep sampling until batches have non-zero-std rewards; assert `async_steps > 1`.
>   - Move reward computation into `accumulate_inference_batches`; now returns `(GenerationResult, Batch, reward_metrics, BatchStatistics)`.
>   - Compute and filter zero-std prompts during accumulation; track filtered counts in `BatchStatistics`.
>   - Update truncated completion masking to filter responses and align all related tensors in-place.
> - **Data structures & APIs**:
>   - Extend `model_utils.Batch` with `decoded_responses` and `scores`.
>   - Add `BatchStatistics` dataclass for prompt/response lengths and filtering stats.
>   - Change `apply_verifiable_reward` and `reward_fn` signatures to accept `ground_truths` and `datasets` instead of `Batch`.
>   - Add `utils.combine_reward_metrics` to aggregate per-prompt reward metrics.
>   - Compute `Args.max_possible_score` for solved/unsolved stats.
> - **Metrics & logging**:
>   - Log aggregated reward metrics and batch filtering stats; include `time/reward` and updated "real/unsolved batch size" ratios.
>   - Evaluation uses `eval_batch.scores` and `eval_batch.decoded_responses` directly.
> - **Training loop**:
>   - Main thread refills prompts based on `num_filtered_prompts`; adjusts episode counting and packing outputs accordingly.
>   - `load_data_from_packing_thread` now returns `num_filtered_prompts`.
> - **Tests**:
>   - Update tests to provide tokenizer/reward_fn and to validate new accumulation outputs and batch sizing.
> - **Scripts**:
>   - Enable `--active_sampling` and set `--async_steps` in debug/integration scripts; minor parameter tweaks.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 9c5ba65edd9336a95f3eabd525d4aefae9d41962. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: open_instruct/grpo_fast.py [hunk 1]
```diff
@@ -1574,11 +1591,50 @@ def accumulate_inference_batches(
...
+                    decoded_responses,
+                    Batch(ground_truths=k_ground_truths, datasets=k_datasets),
+                    result.finish_reasons,
+                    result.request_info,
+                    k_decoded_queries,
+                )
+            )
+
+        # Filter out zero std prompts
+        if filter_zero_std_samples and np.array(scores).std() == 0:
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py]: We can just do `np.std(scores)` without the array!

File: open_instruct/grpo_fast.py [hunk 2]
```diff
@@ -2559,21 +2579,21 @@ def maybe_evaluate(
...
-        eval_scores, eval_reward_metrics = asyncio.run(
-            reward_fn(
-                eval_result.responses,
-                eval_decoded_responses,
-                eval_batch if eval_batch else Batch(queries=[], ground_truths=[], datasets=[], indices=None),
-                eval_result.finish_reasons,
-                eval_result.request_info,
-            )
-        )
+        # eval_decoded_responses = tokenizer.batch_decode(eval_result.responses, skip_special_tokens=True)
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py]: Why comment these out?

File: open_instruct/test_grpo_fast.py [hunk 3]
```diff
@@ -198,6 +198,24 @@ def create_mock_result(self, dataset_index, epoch_number, num_samples_per_prompt
...
+        # Set up dummy reward fn
+        def reward_fn(
+            responses: list[torch.Tensor],
+            decoded_responses: list[str],
+            batch,
+            finish_reasons: list[str],
+            infos: list[list[int]],
+            queries: list[str] | None = None,
+        ) -> list[float]:
+            return [1.0 for _ in range(len(responses))]
```
[the remark this answers is not in the document]
Comment (mnoukhov) [reply open_instruct/test_grpo_fast.py]: fixed

File: open_instruct/grpo_fast.py [hunk 4]
```diff
@@ -1994,30 +1991,37 @@ def data_preparation_thread(
...
-                "packed_ratio": len(packed_sequences.query_responses) / len(responses) if len(responses) > 0 else 0,
-                "val/all_zero_reward_groups": all_zero_groups,
-                "val/all_zero_reward_groups_ratio": all_zero_groups_ratio,
-                "val/total_reward_groups": total_groups,
+                "scores": scores.mean(),
+                "real_batch_size_ratio": real_num_responses / expected_num_responses,
+                "unsolved_batch_size_ratio": solved_num_responses / real_num_responses,
+                "packed_ratio": len(packed_sequences.query_responses) / real_num_responses,
+                "val/solve_rate_hist": None,
+                "val/total_reward_groups": real_num_responses / args.num_samples_per_prompt_rollout,
```
[the remark this answers is not in the document]
Comment (mnoukhov) [reply open_instruct/grpo_fast.py:2010]: this issue is present in the current code as well, we can fix it but I would leave that to future PRs

File: open_instruct/grpo_fast.py [hunk 5]
```diff
@@ -2933,20 +2952,32 @@ def health_check_fn():
...
-        split_and_insert_batch(
-            batch,
-            iter_dataloader.epoch_number,
-            training_step,
-            pending_queries_map,
-            param_prompt_Q,
-            generation_configs["train"],
-            is_eval=False,
-        )
+        num_prompts_to_refill += args.num_unique_prompts_rollout + num_filtered_prompts
```
[the remark this answers is not in the document]
Comment (mnoukhov) [reply open_instruct/grpo_fast.py:2953]: moved until after collated_data is None

File: open_instruct/grpo_fast.py [hunk 6]
```diff
@@ -1839,118 +1912,32 @@ def data_preparation_thread(
...
-            non_zero_std_mask = scores_per_prompt.std(axis=-1) != 0
-            real_batch_size_ratio = non_zero_std_mask.sum() * args.num_samples_per_prompt_rollout / len(scores)
-            expanded_mask = np.repeat(non_zero_std_mask, args.num_samples_per_prompt_rollout)
-            non_zero_gradient_index = np.where(expanded_mask)[0]
-
-            # Log zero-gradient filtering statistics
-            num_zero_std_prompts = (~non_zero_std_mask).sum()
-            num_filtered_responses = len(scores) - len(non_zero_gradient_index)
-            if num_filtered_responses > 0:
+        if args.mask_truncated_completions:
```
[the remark this answers is not in the document]
Comment (mnoukhov) [reply open_instruct/grpo_fast.py:1911]: fine ugh, this isn't an issue from this PR but I fixed it

File: open_instruct/grpo_fast.py [hunk 7]
```diff
@@ -1994,30 +1980,37 @@ def data_preparation_thread(
...
             # Handle empty responses case
             # in this case, we won't log metrics, so it should be fine.
             metrics = {}
+            logger.warning(f"No responses in batch {training_step}.")
         else:
-            sequence_lengths = np.array([len(response) for response in responses])
+            real_num_responses = len(result.responses)
+            expected_num_responses = args.num_samples_per_prompt_rollout * args.num_unique_prompts_rollout
+
+            solved_num_responses = (scores < args.max_possible_score).sum()
```
[the remark this answers is not in the document]
Comment (mnoukhov) [reply open_instruct/grpo_fast.py]: fixed

File: open_instruct/grpo_fast.py [hunk 8]
```diff
@@ -1658,11 +1694,52 @@ def accumulate_inference_batches(
 
         query, ground_truth, dataset, raw_query = pending_queries_map.pop(result.dataset_index)
 
+        for i in range(len(result.finish_reasons)):
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py:1699]:
```suggestion
        # TODO(finbarrtimbers): Move this to LLMRayActor.
        for i in range(len(result.finish_reasons)):
```

File: open_instruct/grpo_fast.py [hunk 9]
```diff
@@ -1658,11 +1694,52 @@ def accumulate_inference_batches(
...
+            if result.finish_reasons[i] == "stop" and len(result.responses[i]) == 0:
+                result.responses[i].append(tokenizer.eos_token_id)
+                result.masks[i].append(1)
+                result.logprobs[i].append(float("nan"))
+
+        decoded_responses = tokenizer.batch_decode(result.responses, skip_special_tokens=True)
+
+        k_queries = [query for _ in range(generation_config.n)]
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py]:
```suggestion
        # TODO(finbarrtimbers): Make PendingQueriesMap.pop return a Batch, and add a Batch.repeat method.
        k_queries = [query for _ in range(generation_config.n)]
```

File: open_instruct/grpo_fast.py [hunk 10]
```diff
@@ -1658,11 +1694,52 @@ def accumulate_inference_batches(
...
+        k_ground_truths = [ground_truth for _ in range(generation_config.n)]
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py]: Can you use the `repeat_each` function here? like we do in packing!

File: open_instruct/grpo_fast.py [hunk 11]
```diff
@@ -1658,11 +1694,52 @@ def accumulate_inference_batches(
...
+        k_datasets = [dataset for _ in range(generation_config.n)]
+        k_raw_queries = [raw_query for _ in range(generation_config.n)]
+
+        # with Timer("💰 [Data Preparation Thread] Calculating rewards and advantages"):
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py]: Delete it!

File: open_instruct/grpo_fast.py [hunk 12]
```diff
@@ -1750,9 +1827,26 @@ def accumulate_inference_batches(
...
+    batch_stats = BatchStatistics(
+        prompt_lengths=prompt_lengths,
+        response_lengths=response_lengths,
+        filtered_prompts=total_filtered_prompts,
+        filtered_prompts_zero=filtered_prompt_zero,
+        filtered_prompts_solved=filtered_prompt_solved,
+        filtered_prompts_nonzero=filtered_prompt_nonzero,
+    )
+    logging.info(
+        f"[Data Preparation Thread] Calculating rewards took {combined_reward_metrics.get('time/reward', 0)} seconds"
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py]: "time/reward" should always be in `combined_reward_metrics`, right? Let's remove the `.get`.

**Review by finbarrtimbers — CHANGES_REQUESTED** [review]

File: open_instruct/grpo_fast.py (hunk 12 above)
Comment (mnoukhov) [reply open_instruct/grpo_fast.py]: it isn't guaranteed to be there for tests, should I add it to the test reward_fn too?

Comment (finbarrtimbers) [reply open_instruct/grpo_fast.py]: yes please!

File: open_instruct/grpo_fast.py (hunk 2 above)
Comment (mnoukhov) [reply open_instruct/grpo_fast.py]: this should be deleted not commented out, it was deleted in a later commit, we're doing all of this stuff in accumulate_inference_batches

File: open_instruct/grpo_fast.py [hunk 13]
```diff
@@ -2362,12 +2346,21 @@ def load_data_from_packing_thread(
...
-    return collated_data, data_thread_metrics, num_total_tokens, num_step_tokens, prompt_lengths, response_lengths
+    return (
+        collated_data,
+        data_thread_metrics,
+        num_total_tokens,
+        num_step_tokens,
+        prompt_lengths,
+        response_lengths,
+        num_filtered_prompts,
+    )
```
[the remark this answers is not in the document]
Comment (mnoukhov) [reply open_instruct/grpo_fast.py:2368]: fixed

File: open_instruct/grpo_fast.py (hunk 12 above)
Comment (mnoukhov) [reply open_instruct/grpo_fast.py]: done!

File: open_instruct/grpo_fast.py (hunk 10 above)
Comment (mnoukhov) [reply open_instruct/grpo_fast.py]: done!

File: open_instruct/grpo_fast.py (hunk 1 above)
Comment (mnoukhov) [reply open_instruct/grpo_fast.py]: done!

File: open_instruct/grpo_fast.py (hunk 11 above)
Comment (mnoukhov) [reply open_instruct/grpo_fast.py]: done!

File: open_instruct/grpo_fast.py [hunk 14]
```diff
@@ -1750,9 +1830,26 @@ def accumulate_inference_batches(
...
+    batch_stats = BatchStatistics(
+        prompt_lengths=prompt_lengths,
+        response_lengths=response_lengths,
+        filtered_prompts=total_filtered_prompts,
+        filtered_prompts_zero=filtered_prompt_zero,
+        filtered_prompts_solved=filtered_prompt_solved,
+        filtered_prompts_nonzero=filtered_prompt_nonzero,
+    )
+    logging.info(
+        f"[Data Preparation Thread] Calculating rewards took {combined_reward_metrics['time/reward']} seconds"
```
[the remark this answers is not in the document]
Comment (mnoukhov) [reply open_instruct/grpo_fast.py:1851]: Nope, we're all good

**Review by finbarrtimbers — APPROVED** [review]

File: open_instruct/grpo_fast.py [hunk 15]
```diff
@@ -2933,20 +2940,31 @@ def health_check_fn():
...
+            collated_data,
+            data_thread_metrics,
+            num_total_tokens,
+            num_step_tokens,
+            prompt_lengths,
+            response_lengths,
+            num_filtered_prompts,
+        ) = load_data_from_packing_thread(packed_sequences_Q, num_total_tokens, stop_event, health_check_fn)
+
+        num_prompts_to_refill += args.num_unique_prompts_rollout + num_filtered_prompts
```
[the remark this answers is not in the document]
Comment (mnoukhov) [reply open_instruct/grpo_fast.py:2953]: you're wrong, the loop below subtracts `num_unique_prompts_rollout` every loop and will do multiple loops if necessary
